File: 3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic =  collections.defaultdict(int)
        l = 0
        n = len(s)
        res = 0

        for r in range(n):
            dic[s[r]] += 1
            while dic[s[r]] > 1:
                dic[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)

        return res

File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0;
        int n = s.size();
        int right = 0;
        int ans = 0;
        unordered_set<char> char_set;
        while(right < n){
            while(char_set.find(s[right]) != char_set.end()){
                char_set.erase(s[left]);
                left++;
            }
            char_set.insert(s[right]);
            ans = max(ans, right - left + 1);
            right++;
        }
        return ans;
    }
};